### PR TITLE
[sabakan] improve sabactl machines get

### PR DESCRIPTION
**Pull request by yamatcha**
Signed-off-by: yamatcha <soju-yamashita@cybozu.co.jp>

### CODE REVIEW

**Comment by zoetrope** [comment]
Please add test cases.
https://github.com/cybozu-go/sabakan/blob/main/e2e/sabactl_test.go

File: models/etcd/index.go [hunk 1]
```diff
@@ -152,8 +153,10 @@ func (mi *machinesIndex) query(q sabakan.Query) []string {
 
 	res := make(map[string]struct{})
 
-	for _, serial := range mi.Rack[q.Rack()] {
-		res[serial] = struct{}{}
+	for _, rack := range strings.Split(q.Rack(), ",") {
```
Comment (zoetrope) [inline models/etcd/index.go:156]: Why can we only specify multiple racks?
We want to specify other options as well.

The original issue is just an example of a rack.

File: pkg/sabactl/cmd/machines.go [hunk 2]
```diff
@@ -191,19 +209,32 @@ var machinesSetRetireDateCmd = &cobra.Command{
...
-		"serial":   "Serial name",
-		"rack":     "Rack name",
-		"role":     "Role name",
-		"labels":   "Label name and value (--labels key=val,...)",
-		"ipv4":     "IPv4 address",
-		"ipv6":     "IPv6 address",
-		"bmc-type": "BMC type",
-		"state":    "State",
+		"serial":           "Serial name",
+		"rack":             "Rack name",
```
Comment (zoetrope) [inline pkg/sabactl/cmd/machines.go:213]: Options that allow multiple specifications should be named in the plural form like `racks`.
You should also add an explanation ("Rack names (--racks 1,2,3...)").

If you don't want to break compatibility, you can provide two options, like `rack` and `racks`.

File: models/etcd/machine.go [hunk 3]
```diff
@@ -258,8 +258,9 @@ RETRY:
 func (d *driver) machineQuery(ctx context.Context, q sabakan.Query) ([]*sabakan.Machine, error) {
 	var serials []string
 
+	qRemovedWithout := q.RemoveWithout()
```
Comment (zoetrope) [inline models/etcd/machine.go:261]: I could not read the intent of this code.
Does "without" mean the same as "empty"?

File: pkg/sabactl/cmd/machines.go [hunk 4]
```diff
@@ -40,9 +45,22 @@ var machinesGetCmd = &cobra.Command{
 			if err != nil {
 				return err
 			}
-			e := json.NewEncoder(cmd.OutOrStdout())
-			e.SetIndent("", "  ")
-			return e.Encode(ms)
+			if ok && *outputFormat == "simple" {
```
Comment (zoetrope) [inline pkg/sabactl/cmd/machines.go:48]: I prefer to separate `output` option into a separate PR.
However, in this case it is not a problem.

File: models/etcd/machine.go (hunk 3 above)
Comment (zoetrope) [reply models/etcd/machine.go:261]: I understood.
But I think this code is not readable.
Please rewrite this code or add a comment.

File: query.go [hunk 5]
```diff
@@ -111,3 +186,53 @@ func (q Query) IsEmpty() bool {
...
+		}
+	}
+	return removed
+}
+
+// Valid returns true if query isn't conflicted
+func (q Query) Valid() bool {
+	hasWithoutSerial := q["without-serial"]
+	if hasSerial := q["serial"]; len(hasSerial) > 0 && len(hasWithoutSerial) > 0 {
+		return false
```
Comment (zoetrope) [inline query.go:207]: Can't users find out why it failed?

File: pkg/sabactl/cmd/machines.go (hunk 2 above)
Comment (zoetrope) [reply pkg/sabactl/cmd/machines.go:213]: Let's discuss if we can break compatibility.

**Review by zoetrope — CHANGES_REQUESTED** [review]

**Comment by kmdkuk** [comment]
I checked for now!
I thought it was good, except for the part that already pointed out about option names and the ability to specify multiple values.
👍

File: query.go (hunk 5 above)
Comment (ysksuzuki) [reply query.go:277]: Sabakan returns `"'with' and 'without' options about the same things are specified."` when users specifies bothe with and without.

**Review by zoetrope — APPROVED** [review]
LGTM

**Pull request merged**